### DISCUSSION TITLE
Chapter progression and "chapter select" logic

### DISCUSSION THREAD

**Discussion by ViMaSter** [Studio-Lovelies/GG-JointJustice-Unity#170] [Ideas]
### Context
This isn't really an actionable item yet (hence no GitHub Issue), but just me extracting a discussion between @PhoebeMitchell and me that came up during https://github.com/Studio-Lovelies/GG-JointJustice-Unity/pull/146#discussion_r793246294. It was out of the scope for the PR but an important technical decision nevertheless, so I figured a thread like this to continue the discussion and get everyone involved makes sense.

This came up while [removing the `_nextSceneName` from the `StoryController` (soon to be `NarrativeScriptPlaylist`)](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/pull/146#discussion_r793246294):

@ViMaSter:
> Nit: I'm unsure if this field and it's associated logic are still relevant after this PR. This field's value is used if we "run out of lines" [inside an Ink narrative script] to execute and I would make the argument, this is something we should throw and abort execution for instead; it's actually an unexpected state:
>
> Generally, the game's flow is planned to be completely defined via the `.ink`-scripts; thereby running out of lines _shouldn't_ really happen. If we encounter the final dialogue for the current state of the game, we should instead transition to credits (or the main menu for the time being) and do so via the `&SCENE:`-action, to give that power to writers instead of limiting it to people working in the editor only. Doing so, the Unity scene is left before "no lines" are left inside the narrative script.
>
> Does that approach make sense to you too? Maybe there's also another case where this could prove useful I've missed, so if I did, do let me know!

@PhoebeMitchell:
> I agree, and it also doesn't make sense for the one scene for all scripts approach we're going for in this PR. Are you suggesting having, for example, a &LOAD_SCRIPT action at the end of every narrative script? This would require a way of either reloading a scene and injecting the script to play (which I think needs to be done at some point anyway), or destroying the assets for the current scene and loading new ones (or not destroying assets that need to be used in the new script if we want to be SUPER efficient). Maybe this should be its own issue and PR as I'm not sure how much time I will be able to dedicate at the moment with university.

Picking up the discussion from here...

---

Now to start, a quick brain dump of how I was envisioning things when I wrote that comment: The `NarrativeScriptPlaylist` would be a big piece to the chapter/scene management puzzle.

In my mind, there's a `Game` scene, which is basically the `Inky - Test Scene` we have right now, but the `NarrativeScriptPlaylist` contains **all** scripts; from the first tutorial case the player sees when he starts a new game, to the final chapter of the game right before the credits.

When starting a new game...
- **I. 1.**: this scene gets loaded
- **I. 2.**: the story controller starts with the first script available
- **I. 3.**: the logic implemented by you in #146 is ran for **only that script**
- **I. 4.**: the game runs until the script is over
- **I. 5.**: the story controller moves to the next script and goes to 3 if there is one
- **I. 6.**: the final ink script uses the `&SCENE:Credits` line to abort that loop and exit the `Game` scene before the `NarrativeScriptPlaylist` runs out of scripts to parse

If we want to continue from a chapter...
- **II. 1.**: the game (somehow?) can map a chapter selection to a `.ink`-script
- **II. 2.**: the `Game` scene gets loaded **BUT...**
- **II. 3.**: we remove scripts from the top of the list, until we encounter the "chapter" we've selected
- **II. 4.**: then we go to `I. 1.`

Now, the `&LOAD_SCRIPT` you mentioned actually solves a super important problem: In my mind, that `Game`-scene with the full `NarrativeScriptPlaylist` had to be pre-built inside the Unity editor.  
With your suggestion, we could actually have those "links" between scenes be created at runtime via `.ink`-scripts as well, which effectively makes this `Game` scene a "C# method in disguise" called `StartGameFrom(scriptFile)`:
Instead of removing entries from a pre-setup list, we would...
1. take any `.ink`-file as entry point
  1. 2. `New Game` would select the first narrative file
  1. 2. `Select Chapter` is just a map to any other, later file
2. search for `&LOAD_SCRIPT` actions and queue them as next file to parse
3. do this until we encounter no more `&LOAD_SCRIPT`-action lines
4. we have a populated `NarrativeScriptPlaylist` and can start the game

This feels super robust and modular to me and is actually already detailed enough to be actionable.  
Happy to hear thoughts from everyone else on this!

**Comment by PhoebeMitchell** [comment]
I like the method you detailed using the &LOAD_SCENE action. Could be subject to infinite loops if two narrative scripts are accidentally told to load each other, but not exactly a difficult problem to solve. I don't have much to add on that front.

Maybe a separate issue but I think credits should have a dedicated scene, and shouldn't be handled with a BGScene as the action &SCENE:Credits implies.

>  the game (somehow?) can map a chapter selection to a .ink-script

The age old question of how to pass data to a Unity scene when loading it. The methods I know are saving/load to disk, using a static variable, or DontDestroyOnLoad (which usually requires a static variable or Find method to access relevant data in the new scene anyway). My idea was to have a static variable on NarrativeScriptPlaylist that can be set in the main menu scene, which can then be accessed in the Awake method when the next scene starts. Narrative scripts could be assigned to chapter select buttons, which set the static variable to their assigned narrative script, and load the Game scene.

**Reply by ViMaSter** [reply]
> Could be subject to infinite loops if two narrative scripts are accidentally told to load each other, but not exactly a difficult problem to solve.

Checking for that is actually a good idea; we could just add an editor script that we can call from our GitHub Actions so that we're watching for circular references on each build.

> The age old question of how to pass data to a Unity scene when loading it. 

I actually think something along the lines of
1. `LoadScene("game")`
2. `GameObject.FindObjectOfType<NarrativeScriptPlaylist>().`... -> do whatever we need to do to update the playlist to start at the selected chapter
3.  `UnloadScene("menu")`

should already work, as (as far as I know) `Start()` of the `NarrativeScriptPlaylist` is only called on the next frame after calling `LoadScene()` and not directly after or while calling `LoadScene()` (unlike `Awake()` which gets called right after a component is added or enabled). If we manage to update the playlist before `Start()` is called, nothing should be initialized and we should be good to go.

**Reply by PhoebeMitchell** [reply]
Not a fan of find methods, even if the performance overhead would be minimal. FindObjectOfType is admittedly not as bad as other find methods due to having no string dependency, but would still like a more direct solution that doesn't require checking every component in a scene.

**Reply by ViMaSter** [reply]
> Not a fan of find methods, even if the performance overhead would be minimal.

Especially given it's only ran once, not every frame or in a fixed interval. That said, loading a scene might already take a few seconds, so trying to keep overhead minimal is definitely a good thing. And I agree with `Find`s generally feeling a bit too vague.

Spinning off of your static field idea: How about making `NarrativeScriptPlaylist` a singleton component? On `Awake()` it assigns itself to `public static Instance {get; private set;}`, throws if it exists already (since I don't think having multiple playlists running in parallel is a good idea) and things like the script handling the chapter selection can then use `NarrativeScriptPlaylist.Instance.` to access it directly.

Does that sounds like a good compromise?

**Reply by PhoebeMitchell** [reply]
Making NarrativeScriptPlaylist a singleton would imply anything is able to access it. This may be a good thing because lots of things need to access it anyway and it may clean up the reference spaghetti have currently have.

However its one of those things I'm always told to avoid (along with public fields for instance). That along with the that moving away from singletons was one of the reasons we decided to start from scratch instead of using Jape's build.

That said I recognise I often go down the path of following these "best practices" without fully understanding why they're bad. At the end of the day I'd like to be able to show this project off to someone and defend our decisions. So if we decide on a singleton I'd like to be able to say why it was better than any other method.

Having a private field on NarrativeScriptPlaylist would enforce the idea of not everything having access to it. An instance of it can exist in whatever scene has the scene select logic, the static field can be set via property, then NarrativeScriptPlaylist is free to use that in the game scene. I guess a drawback of that is if we forget to set the field to null after getting the script from it, and causing some unexpected error elsewhere.

**Reply by ViMaSter** [reply]
Generally I also try to avoid `static` and Singletons.  
- (Lazily initialized) Singletons have this nasty side-effect of a runtime-specific point of initialization; it's not possible to just deduct the point of creation just by looking at the code, as it's only set up when it's accessed for the first time. (Effectively any call to `.Instance` is both a potential call to the constructor.) Especially in bulk, this can make code hard to read, cause spikes of computation at unexpected times and (in the worst case) cause dependency-cycles; all of which aren't obvious until the code is actually run.  
- In regards to `static`: As OOP is generally less painful when things are encapsulated, having a field, property or even entire component that is "accessible from anywhere, all the time", is basically the antithesis of it.
- About `DontDestroyOnLoad`: I find it super hard to wrap my head around this; it's a simple concept in theory, in practice it never solved a problem the way I wanted it to. That aside, it's also something Unity-specific which I try to keep to a minimum, unless we're talking to Engine implementations for which there's no cross-platform equivalent inside .NET Framework already.

Just did a bit of googling around (`unity load level with parameters`) and I basically keep running into the same problem and solutions:
https://forum.unity.com/threads/unity-beginner-loadlevel-with-arguments.180925/
https://stackoverflow.com/questions/42393259/load-scene-with-param-variable-unity

From the top of my head I can't really come up with anything aside from a static field or a one-time `Find...()`-call (which I think I still prefer over using anything `static`). Definitely open for any other suggestions!

**Reply by PhoebeMitchell** [reply]
I guess as Find methods can find any object without a reference they achieve a similar thing to static fields, just with extra overhead.

> having a field, property or even entire component that is "accessible from anywhere, all the time", is basically the antithesis of it.

Unsure if this is in reference to my idea, but in my implementation the field in question would be private. It would only be accessible by NarrativeScriptPlaylist, and other objects would require a reference to an instance of NarrativeScriptPlaylist to be able to set it by property.

Alternatively we create a wrapper for scene manager that allows parameters, though that would use even more static stuff. I don't think static is inherently bad, we should just avoid using it to allow global access.

**Reply by PhoebeMitchell** [reply]
Oooor we make chapter selection part of the Game scene and don't worry about any of this :p

**Reply by ViMaSter** [reply]
> Oooor we make chapter selection part of the Game scene and don't worry about any of this :p

I think I'd rather worry about how to pass data from scene to scene, than having a "Monoscene" be the entire game. : D  
Especially from a testability standpoint, if we decide to add things like mini-games or other scenes with limited/new input logic (like the Credits), I'd prefer if we can launch them on their own.

Writing that out, and since I'm still [a fan of data-binding](https://github.com/ViMaSter/unity-databinding), that would also be an option: 
 
Aside from scene-specific things like "contents of the `AppearingDialogueController`" (or instead of using it for that), we can have it be a global JSON-blob where we persist things like our audio and video settings or selections that need to be persisted across scenes.
With that, every new AudioSource can just check in-memory which volume it's set to.  
Similarly the `NarrativeScriptPlaylist` can load info about which chapter to start from from there:
1. The chapter select button would persist the target chapter under `global.chapterSelect.startChapter = "chapter5"`
2. We load the game scene
3. The `NarrativeScriptPlaylist`  checks if a `global.chapterSelect.startChapter` key is set
    1. If so, it skips to the specified chapter
    2. if not, it starts from the beginning

**Reply by PhoebeMitchell** [reply]
NotExplosive vouches for the monoscene idea. However despite me suggesting it, it don't think I'd like to go down that route either, mainly because loading scenes would be a lot easier than constructing them manually, and I'd very much like to get rid of the standard inclusion prefab. 

In regard to the JSON databinding, I don't exactly understand the advantage of that as opposed to using static fields in C#. Seems like they both allow you to access data from anywhere, but one is quite complex and I don't understand it. I've been meaning to take a look at the databinding package you linked to try and understand what it's doing, but heh, time and stuff :p That said, if it persists data between opening and closing the game that could be very useful for audio settings etc, but the scene to load isn't really something that needs to persist between opening and closing the game.

If you really don't want private fields, I'd prefer the Find method approach, mainly due to simplicity. I feel we could discuss this forever, but I'd like to get scene loading implemented soonish if possible.

**Reply by ViMaSter** [reply]
> NotExplosive vouches for the monoscene idea. However despite me suggesting it, it don't think I'd like to go down that route either, mainly because loading scenes would be a lot easier than constructing them manually

I actually still really like that idea of at least keeping all narrative scripts limited to one Unity scene and loading everything based on `.ink`-files. But anything that breaks out of the "cross examination or linear narrative"-flow (things like credits, main menu, chapter select, settings or [things like the "evidence gathering"-sections we might want to add later on](https://www.youtube.com/watch?v=jmAOL8MLgOY&t=10m11s)) I could definitely still see receiving their own scene.

> and I'd very much like to get rid of the standard inclusion prefab.

This actually confused me a bit: If we don't want to move forward with "one scene for all narrative scripts" and have multiple scenes instead, wouldn't we have to utilize prefabs again? Having only one scene would reduce the need for it more, at least in my mind.

> If you really don't want private fields, I'd prefer the Find method approach, mainly due to simplicity. I feel we could discuss this forever, but I'd like to get scene loading implemented soonish if possible.

Yeah, I think I prefer the `.Find()` method approach. Using an instance of a class, to modify a private static field would result in code that is unintentionally "hiding" the logic underneath: Every class interfacing with that fields would never directly access the static field, but call a method on an instance, which then would access the static field in its stead. And generally I wouldn't expect an instanced method to modify static data. If the field was public, at least we'd directly see scripts making changes to a static field.

---

Also moved the questions for the databinding stuff [to a separate thread](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/discussions/195#discussioncomment-2192241), so this thread can be focused on chapter progression. If we actually decide to move forward with that idea in general, we can always refactor things later on. Hope that's okay with you too. 👀

**Reply by PhoebeMitchell** [reply]
I'm still on board with having one scene for trials. I meant I don't like the idea of the main menu, credits, other stuff also being in that scene, they should have their own dedicated scenes. The standard inclusion prefab was to allow us to keep multiple trial scenes up to date, but it would be unnecessary if we just had the one scene dedicated to trials, which is what I meant when I said I wanted to get rid of it. Sorry if I worded it confusingly.

Aside from that it seems we agree on using find.

**Reply by ViMaSter** [reply]
Re https://github.com/Studio-Lovelies/GG-JointJustice-Unity/pull/189#discussion_r810609832:
> Why is loading all the scripts and then putting them in a playlist required though, if we aren't using anything in those scripts until we reach them? If resources are only loaded when a new script starts (which I agree should be the case), then we can just load the next script when it's needed.

I think the only reason for it in my mind was that - right now at least - we don't have a way to refer to "the next script" inside scripts themselves (which is probably the cleaner solution?).  
As a possible alternative: Instead of having the playlist set-up in the Unity Editor, we could also just have a `&QUEUE_NEXT_SCRIPT:scriptname`-action (similar to `&SCENE`, just that it stays in the `Game` scene and loads the resources for the next script and sets it as active), which basically does the same thing as the `NarrativeScriptPlaylist` does right now when it encounters the end of a script and loads the next one. This would reduce the need to have a playlist and give the same level of flexibility.

Would something like that work for you?
